doc/tunnel/SAI-Proposal-IP-in-MPLS-Tunnel.md

```diff
@@ -0,0 +1,776 @@
+1. Node 1 programs label stack [5, 3] (with 5 as BoS) and forwards packet to node 2 (PUSH).
+2. Node 1 also programs label 1 for UHP (NEXT)
+3. Node 2 swaps label 3 with label 3 (CONTINUE)
+4. Node 2 also programs label 2 for UHP(NEXT)
+5. Node 3 programs label 3 for UHP (NEXT)
+6. Node 5 programs label 5 for  UHP (NEXT)
+7. Node 5 also programs label stack [1, 2] (with 1 as BoS) and forwards packet to node 3(PUSH)
+
+For a path A-1-4-5-B and B-5-4-1-A
+1. Node 1 programs label stack [201] (with 201 as BoS) and forwards packet to node 4(PUSH)
+2. Node 1 also programs label 101 for UHP (NEXT)
+3. Noose 4 performs PHP for label 101 and label 201 and forward to Node 1 and Node 5 respectively (NEXT)
+4. Node 5 programs label stack [101] (with 101 as BoS) and forwards packet to node 4(PUSH)
+5. Node 5 also programs label 201 for UHP(NEXT)
+
+Below is SAI configuration at each node to achieve this.
+
+## Node Configuration
+
+### Node 1 configuration
+
+#### router interfaces
+
+```
+rif_1_a = {
+    {SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID, vr0},
+    {SAI_ROUTER_INTERFACE_ATTR_TYPE, SAI_ROUTER_INTERFACE_TYPE_VLAN},
+    {SAI_ROUTER_INTERFACE_ATTR_VLAN_ID, vlan_1_a},
+    {SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS, router1_mac},
+}
+rif_1_2 = {
+    {SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID, vr0},
+    {SAI_ROUTER_INTERFACE_ATTR_TYPE, SAI_ROUTER_INTERFACE_TYPE_VLAN},
+    {SAI_ROUTER_INTERFACE_ATTR_VLAN_ID, vlan_1_2},
+    {SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS, router1_mac},
+}
+rif_1_4 = {
+    {SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID, vr0},
+    {SAI_ROUTER_INTERFACE_ATTR_TYPE, SAI_ROUTER_INTERFACE_TYPE_VLAN},
+    {SAI_ROUTER_INTERFACE_ATTR_VLAN_ID, vlan_1_4},
+    {SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS, router1_mac},
+}
+mpls_rif = {
+    {SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID, vr0},
+    {SAI_ROUTER_INTERFACE_ATTR_TYPE, SAI_ROUTER_INTERFACE_TYPE_MPLS_ROUTER},
+}
+```
+
+
+#### fdb entries
+
+```
+fdb_a = {
+    {switch_id, vlan_1_a, mac_a},
+    {
+        {SAI_FDB_ENTRY_ATTR_TYPE, SAI_FDB_ENTRY_TYPE_STATIC},
+        {SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID, port_1_a}
+    }
+}
+fdb_2 = {
+    {switch, vlan_1_2, mac_2},
+    {
+        {SAI_FDB_ENTRY_ATTR_TYPE, SAI_FDB_ENTRY_TYPE_STATIC},
+        {SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID, port_1_2}
+    }
+}
+fdb_4 = {
+    {switch, vlan_1_4, mac_4},
+    {
+        {SAI_FDB_ENTRY_ATTR_TYPE, SAI_FDB_ENTRY_TYPE_STATIC},
+        {SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID, port_1_4}
+    }
+}
+```
+
+#### neighbor entries
+
+```
+neighbor_a = {
+    {switch_id, rif_1_a, 101.101.0.101},
+    {
+        {SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS, mac_a}
+    }
+}
+neighbor_2 = {
+    {switch_id, rif_1_2, 1.1.1.21},
+    {
+        {SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS, mac_2},
+    }
+}
+neighbor_4 = {
+    {switch_id, rif_1_4, 1.1.1.41},
+    {
+        {SAI_NEIGHBOR_ENTRY_ATTR_DST_MAC_ADDRESS, mac_4}
+    }
+}
+```
+
+#### decap tunnel
+
+```
+decap_tunnel = {
```

Comment:
where is this decap_tunnel object id get used?

Reply:
decap tunnel is supposed to be used by all PHP operations